### DISCUSSION TITLE
Vectors in Drasil

### DISCUSSION THREAD

**Discussion by BilalM04** [JacquesCarette/Drasil#3923] [General]
Moving the vector content from #3892 here, so it doesn't get lost.

## Current State of Vectors

There is currently *some* vector support in `drasil-lang`:
  - You can declare vectors: [`vec2D`](https://github.com/JacquesCarette/Drasil/blob/3261db372ce2e9ade85942ab08e83ef834a6ed47/code/drasil-lang/lib/Language/Drasil/Expr/Class.hs#L68-L69), [`rowVec`](https://github.com/JacquesCarette/Drasil/blob/3261db372ce2e9ade85942ab08e83ef834a6ed47/code/drasil-lang/lib/Language/Drasil/Expr/Class.hs#L79-L81), and [`columnVec`](https://github.com/JacquesCarette/Drasil/blob/3261db372ce2e9ade85942ab08e83ef834a6ed47/code/drasil-lang/lib/Language/Drasil/Expr/Class.hs#L83-L85).
  - Support for some vector operations as well: [`dim`](https://github.com/JacquesCarette/Drasil/blob/3261db372ce2e9ade85942ab08e83ef834a6ed47/code/drasil-lang/lib/Language/Drasil/Expr/Class.hs#L163-L164), [`norm`](https://github.com/JacquesCarette/Drasil/blob/3261db372ce2e9ade85942ab08e83ef834a6ed47/code/drasil-lang/lib/Language/Drasil/Expr/Class.hs#L166-L167), [`negVec`](https://github.com/JacquesCarette/Drasil/blob/3261db372ce2e9ade85942ab08e83ef834a6ed47/code/drasil-lang/lib/Language/Drasil/Expr/Class.hs#L169-L170), [`cross`, `vScale`, `vAdd`, `vSub`](https://github.com/JacquesCarette/Drasil/blob/3261db372ce2e9ade85942ab08e83ef834a6ed47/code/drasil-lang/lib/Language/Drasil/Expr/Class.hs#L187-L197), and [`dot product`](https://github.com/JacquesCarette/Drasil/blob/3261db372ce2e9ade85942ab08e83ef834a6ed47/code/drasil-lang/lib/Language/Drasil/Expr/Class.hs#L103-L104).

There also seems to be some support for vectors in GOOL, implementing vectors and their operations. This [vector test file](https://github.com/JacquesCarette/Drasil/blob/main/code/drasil-code/test/VectorTest.hs) shows that some vector features are operational in GOOL.

## My Work

As a start, I changed the [calculation of landing time](https://jacquescarette.github.io/Drasil/examples/projectile/SRS/mdBook/book/SecIMs.html#IM:calOfLandingTime) instance model for projectile to use vectors. I changed it to the following: 

```Haskell
flightDur' = exactDbl 2 $* (vec2D (sy launSpeed $* cos (sy launAngle)) (sy launSpeed $* sin (sy launAngle)) $. vec2D (exactDbl 0) (exactDbl 1)) $/ sy gravitationalAccelConst
```

$${t_{\text{flight}}}=\frac{2\\,\begin{bmatrix}
                                        {v_{\text{launch}}}\\,\cos\left(θ\right)\\
                                        {v_{\text{launch}}}\\,\sin\left(θ\right)
                                        \end{bmatrix}\cdot{}
                                        \begin{bmatrix}
                                        0\\
                                        1
                                         \end{bmatrix}}{g}$$

From #3892, it was suggested to incorporate vector scaling:

```Haskell
flightDur' = ((exactDbl 2 $/ sy gravitationalAccelConst) `vScale` vec2D (sy launSpeed $* cos (sy launAngle)) (sy launSpeed $* sin (sy launAngle))) $. vec2D (exactDbl 0) (exactDbl 1)
```

$${t_{\text{flight}}}=\frac{2}{g} \begin{bmatrix}
                                           {v_{\text{launch}}}\\,\cos\left(θ\right)\\
                                           {v_{\text{launch}}}\\,\sin\left(θ\right)
                                           \end{bmatrix}\cdot{}\begin{bmatrix}
                                                               0\\
                                                               1
                                                               \end{bmatrix}$$

--- 

My approach was to modify one of the instance models using the current vector support. Then, follow the propagation of errors in order to find where it is lacking. I ran into problems at [`convExpr`](https://github.com/JacquesCarette/Drasil/blob/3261db372ce2e9ade85942ab08e83ef834a6ed47/code/drasil-code/lib/Language/Drasil/Code/Imperative/Import.hs#L286-L360), which converts `Expr`s to GOOL values. Here, [column vectors](https://github.com/JacquesCarette/Drasil/blob/3261db372ce2e9ade85942ab08e83ef834a6ed47/code/drasil-code/lib/Language/Drasil/Code/Imperative/Import.hs#L356) and [vector operations](https://github.com/JacquesCarette/Drasil/blob/3261db372ce2e9ade85942ab08e83ef834a6ed47/code/drasil-code/lib/Language/Drasil/Code/Imperative/Import.hs#L460-L472) error out.

From speaking with @B-rando1, my understanding is that GOOL (somewhat?) supports vectors; however, we are lacking the support for converting vectors to GOOL. Additionally `convExpr` converts `Expr`s to `SValue`; however, the [vector operations in GOOL](https://github.com/JacquesCarette/Drasil/blob/3261db372ce2e9ade85942ab08e83ef834a6ed47/code/drasil-gool/lib/Drasil/GOOL/InterfaceCommon.hs#L320-L345) expect vectors to be of type `VSThunk`, which I'm not entirely sure what it is, it has something to do with variables. It doesn't seem to be used for anything in GOOL other than vectors. 

## Question

How should I go about converting vectors and vector operations into GOOL's representation of them?

**Comment by smiths** [comment]
Thank you for the summary @BilalM04.  We now have a better idea of what has been done and what must be done.  The display looks much nice in Projectile with the scalar multiple version of the equation.  However, this appears to just be for the display.  Do you have a variable of type vector in the revised Projectile?

**Reply by BilalM04** [reply]
This was briefly discussed in #3913, I will summarize the main points here.

- The instance models for the display and the generated code are created from the same piece of code. So, changing the expression for one, will automatically change it for the other.
- Initially, I declared the entire expression together in one line; however, now I am extracting the vectors into their own models and variables.

**Comment by BilalM04** [comment]
## Update

In Drasil, Instance Models get converted into Calculation functions. I found there to be some limitations in how the calculation functions are currently implemented. `genCalFunc` takes in a `CodeDefinition`, and based on its definition type, creates the calculation function.

https://github.com/JacquesCarette/Drasil/blob/eb5e8eb9661e10e9922b10569f601255c929de32/code/drasil-code/lib/Language/Drasil/Code/Imperative/Modules.hs#L507-L517

https://github.com/JacquesCarette/Drasil/blob/eb5e8eb9661e10e9922b10569f601255c929de32/code/drasil-code/lib/Language/Drasil/Chunk/CodeDefinition.hs#L13-L14

In the code blocks above, there are two possible definition types, `ODE` and `Definition`.  In the `Definition` case, the equation is converted to GOOL using `convExpr` (who's limitations are mentioned in the OP) and returned in a one-line return statement. For the `ODE` case, the steps for solving an ODE is followed. 

The limitation here is that, since vector operations are done across multiple lines (using for loops), we need to be able to create intermediary variables, perform operations, then return the result, which is not currently supported. There is a `genCalcBlock` helper function, shown below, which has the ability to pattern match on the `CodeExpr` to provide differing behaviours based on the expression (i.e. a `Case` statement). However, doing that with vector operations will be difficult as vectors and vector operations are often mixed in with other operations.

https://github.com/JacquesCarette/Drasil/blob/eb5e8eb9661e10e9922b10569f601255c929de32/code/drasil-code/lib/Language/Drasil/Code/Imperative/Modules.hs#L530-L540

**Reply by BilalM04** [reply]
I was able to generate code for the following Instance Model involving vectors:

![image](https://github.com/user-attachments/assets/aa092ca3-b2ed-42b9-9fdd-61db7114ca68)

Here is the generated Java code:

```Java
public static ArrayList<ArrayList<Double>> func_v_launch_vect(InputParameters inParams) {
    return new ArrayList<ArrayList<Double>>(Arrays.asList(new ArrayList<Double>(Arrays.asList(inParams.v_launch * Math.cos(inParams.theta))), new ArrayList<Double>(Arrays.asList(inParams.v_launch * Math.sin(inParams.theta)))));
}
```

I did have to change the `Space` -> `[CodeType]` mapping for vectors from:

https://github.com/JacquesCarette/Drasil/blob/eb5e8eb9661e10e9922b10569f601255c929de32/code/drasil-code/lib/Language/Drasil/Code/Code.hs#L26

to the following:

```Haskell
spaceToCodeType (S.Vect s) = map (List . List) (spaceToCodeType s)
```

However, doing the above broke the dblpend example. I suspect it is because row vectors / matrices with a single row are implemented with a 1D list, rather than 2D. I am still quite unsure so more investigation is required for that.
